src/app/common/theme-provider.tsx

```diff
@@ -0,0 +1,74 @@
+import {
+  createContext,
+  Dispatch,
+  FC,
+  SetStateAction,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
+
+type ComputedTheme = 'dark' | 'light';
+export type UserSelectedTheme = ComputedTheme | 'system';
+
+export const labelMap = new Map<UserSelectedTheme, string>([
+  ['light', 'Light'],
+  ['dark', 'Dark'],
+  ['system', 'System'],
+]);
+
+export const getThemeLabel = (theme: UserSelectedTheme) => {
+  return labelMap.get(theme);
+};
+
+const ThemeContext = createContext<
+  [ComputedTheme, UserSelectedTheme, Dispatch<SetStateAction<UserSelectedTheme>>]
+>([
+  // These values are not used, but are set to satisfy the context's value type.
+  'light',
+  'system',
+  () => {},
+]);
+
+const getSystemTheme = () =>
+  window.matchMedia('prefers-color-scheme: dark').matches ? 'dark' : 'light';
+
+export const ThemeSwitcherProvider: FC = ({ children }) => {
+  const [userSelectedTheme, setUserSelectedTheme] = useState<UserSelectedTheme>('system');
+  const [theme, setTheme] = useState<ComputedTheme>(getSystemTheme());
```

Comment:
```suggestion
  const [theme, setTheme] = useState<ComputedTheme>(() => getSystemTheme());
```
Prevents this running on every render